### DISCUSSION TITLE
Loading a class which is part of otel extension jar  from Advice class is failing with NoClassDefFoundError

### DISCUSSION THREAD

**Discussion by MadhusudanN** [open-telemetry/opentelemetry-java-instrumentation#9701] [Q&A]
**Background ::**
 I have a spring boot application. I have added custom instrumentation to apply advice when filter method is called as demonstrated in this link :: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/instrumentation/DemoServlet3Instrumentation.java
We know that advice is a static inner class as explained in this document :: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md

**Requirement ::**

1. Provide a default implementation for Advice. (StandardAdvice class)
2. Provide an option to override the default advice(One override method to add custom attributes of choice) (CustomAdvice class)
3. In default implementation (StandardAdvice clas),  since we need static method to apply the Advice annotation,  @Advice.OnMethodEnter(suppress = Throwable.class) , this static method should do the below operations.
               1. Read a system property to get the advice class name (-DinstrumentationAdvice=com.test.incubation.opentelemetry.extension.instrumentation.CustomAdvice)
               2. Load this class using the Class.forName(className)
               3. Get instance of this class using java.lang.reflection API
               4. Invoke the method using   java.lang.reflection API, capture the extra attributes and add them to the baggage and root span
               5. Perform the default operation (Get userid and username from incoming request) and add them to baggage and root span.
               
               
   
Issue ::
Getting the NoClassDefFoundError below ::
Failed to load class due to java.lang.NoClassDefFoundError: com/test/incubation/opentelemetry/extension/instrumentation/CustomAdvice
java.lang.NoClassDefFoundError: com/test/incubation/opentelemetry/extension/instrumentation/CustomAdvice
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:50)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.test.incubation.engineering.springboot.filter.TestFilter2.doFilter(TestFilter2.java:40)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.test.incubation.engineering.springboot.filter.TestFilter1.doFilter(TestFilter1.java:44)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at com.test.cloud.service.security.AgentAuthenticationFilter.doFilter(AgentAuthenticationFilter.java:107)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)

Caused by: java.lang.ClassNotFoundException: com.test.incubation.opentelemetry.extension.instrumentation.CustomAdvice
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:592)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)

Approaches tried ::
Tried to load the class "com/test/incubation/opentelemetry/extension/instrumentation/CustomAdvice" using different class loaders ::
Class.forName()  -- current classloader
ClassLoader.getSystemClassLoader().loadClass  -- Using System class loader
Thread.currentThread().getContextClassLoader().loadClass

The above approaches did not help and ended up with ClassNotFoundException only.

Please find the classes below ::

Interface BaseAdvice class ::
public interface BaseAdvice {

	public default Map<String, String> getAttributesMap(ServletRequest request) {
		return Collections.EMPTY_MAP;
	}

}

StandardAdvice class ::

package com.test.incubation.opentelemetry.extension.instrumentation;

public class StandardAdvice implements BaseAdvice {
	

	@Advice.OnMethodEnter(suppress = Throwable.class)
	public static void onEnter(@Advice.Argument(value = 0) ServletRequest request) {

		
//			System.out.println("Inline is set to true");
			System.out.println("Running Advice");
			Map<String, String> attributesMap = new HashMap<>();
			Class<?> adviceClazz = null;

			
			
			try {

				System.out.println("Check if Advice Type is configured");						
				
				String className = System.getProperty("instrumentationAdvice");
				System.out.println("Loading class :: " + className);
				adviceClazz = Thread.currentThread().getContextClassLoader().loadClass(className);
				//adviceClazz = ClassLoader.getSystemClassLoader().loadClass(className);
//				adviceClazz = Class.forName(className);
				Constructor<?> constructor = adviceClazz.getConstructor();
				Object instance = constructor.newInstance();

				String methodName = "getAttributesMap";
				Method method = adviceClazz.getMethod(methodName, ServletRequest.class);

				Map<String, String> additionalAttrMap = (Map<String, String>) method.invoke(instance, request);

				attributesMap.putAll(additionalAttrMap);

			} catch (Throwable e) {

				System.out.println("Failed to load class due to " + e);
				e.printStackTrace();
			}
			 
			 

//			System.out.println(adviceClazz.getSimpleName() + "::  Running advice");

			/*
			 * System.out.println( adviceClazz.getSimpleName() +
			 * " :: Attributes Map override contents are " + attributesMap.toString());
			 */
			
			
			if (RequestContext.getRequestType() != null) { // already authenticated
				System.out.println(
						/* adviceClazz.getSimpleName()  + */  "::  Already authenticated using -"
						+ RequestContext.getRequestType());

				// Get the user name and user_id from Request
				if (RequestContext.getUserInfo() != null) {

					UserInfo userInfo = (UserInfo) RequestContext.getUserInfo();
					String userId = userInfo.getId();
					String userName = userInfo.getName()
					attributesMap.put("user.id", userId);
					attributesMap.put("user.name", userName);

					System.out.println(String.format(
							/* adviceClazz.getSimpleName() + */ ":: User Id is %s and User name is %s",
							userName, userId));
					
					System.out.println("Attributes added to baggage are " + attributesMap);
					
				}

				
				
				
				
				BaggageBuilder baggageBuilder = Baggage.current().toBuilder();

				for (String key : attributesMap.keySet()) {
					baggageBuilder = baggageBuilder.put(key, attributesMap.get(key));
				}
				
				baggageBuilder.build().storeInContext(Context.current()).makeCurrent();
				
				Span root = LocalRootSpan.current();

				for (String key : attributesMap.keySet()) {
					root = root.setAttribute(key, attributesMap.get(key));
				}

			}
			}
		

	}
	
CustomAdvice class ::

package com.test.incubation.opentelemetry.extension.instrumentation;

public class CustomAdvice extends StandardAdvice {
	@Override
	public Map<String, String> getAttributesMap(ServletRequest request) {
		Map<String, String> attributesMap = new HashMap<>();
		HttpServletRequest httpRequest = (HttpServletRequest) request;
		attributesMap.put("custom-header", httpRequest.getHeader("custom-header"));
		System.out.println("Custom header added is " + attributesMap.get("custom-header"));
		return attributesMap;
	}

}

The arguements passed to the application are shown below ::
set "CATALINA_OPTS=%CATALINA_OPTS%-Dotel.service.name=TestSvc -javaagent:**C:\Open_Telemetry\opentelemetry-javaagent.jar** -Dotel.exporter.otlp.headers="Authorization=Bearer <abcd>" -Dotel.exporter.otlp.endpoint=https://<APM-endpoint-url>:443 -Dotel.traces.exporter=otlp -Dotel.logs.exporter=otlp -Xverify:none -D**otel.javaagent.extensions=C:\eclipse-workspace\opentelemetry-java-instrumentation-extension-demo-1.0-all.jar** **-DinstrumentationAdvice=com.test.incubation.opentelemetry.extension.instrumentation.CustomAdvice**"

Note ::
The above app is deployed as war in tomcat 8.5.84 version.
OS :: Windows 11 

Can someone please look into this and please suggest how can I proceed further on this.

**Comment by laurit** [comment]
See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md#tell-the-agent-which-classes-are-a-part-of-the-instrumentation-by-overriding-the-ishelperclass-method You probably need to override `isHelperClass` so that the agent would know that `com.test.incubation.opentelemetry.extension.instrumentation.CustomAdvice` should be injected into the application class loader.
